EC-84: Restructure eZ Commerce doc, part 2 (#2)
ezsystems/commerce-documentation: docs/guide/basket/basket_features/storing_parts_of_the_product_in_the_basket_line.md

# Storing parts of the product in the basket line

The `CatalogElementSerializeService` service is used for the serializing and unserializing of catalog element.
It implements the `SerializeServiceInterface`.

> This feature ensures that the basket always contains information about the product which should be ordered. If the customer is sending an order, the shop can ensure sure that all important product data for the order is present even if the product was meanwhile removed from the product catalog. 

Reviewer: I am thinking whether the source actually means:
``` suggestion
This feature ensures that the basket always contains information about the products that are being ordered. When the customer sends an order, the shop is able to ensure that all important product data for the order is present, even if the product was meanwhile removed from the product catalog.
```

> This feature ensures that the basket always contains information about the product which should be ordered. If the customer is sending an order, the shop can ensure sure that all important product data for the order is present even if the product was meanwhile removed from the product catalog. 

Reviewer 2: ```suggestion
This feature ensures that the basket always contains information about the product which should be ordered. When the customer sends an order, the shop can verify if all relevant product data for the order is present, even if the product was removed from the product catalog. 
```

The parameter `refreshCatalogElementAfter` defines how long the catalog element will be used from the basket line until it is refreshed from the catalog.

``` yaml
ses_basket.default.refreshCatalogElementAfter: 1 hours
```

The service is using all attributes of the catalog element, that are set in the configuration:

``` yaml

> #defines catalog element attributes that should be stored in a basket line

Reviewer: "should be" or just "are"?

Reviewer 2: code comment

siso_basket.default.stored_catalog_element_attributes:
    Silversolutions\Bundle\EshopBundle\Product\OrderableProductNode:
        baseAttributes: 
            - customerPrice
            - ean
            - identifier
            # ...
        dataMapAttributes: []
    Silversolutions\Bundle\EshopBundle\Product\OrderableVariantNode:
        baseAttributes: 
            - customerPrice
            - ean
            - identifier
            # ...
        dataMapAttributes: [] 
```

!!! note

    By default, all base attributes are set in the configuration and the dataMap attributes are empty.

>     You can override this configuration, but make sure you have specified at least all required base attributes,

Reviewer: "specified all the required" - at least is redundant here

>     You can override this configuration, but make sure you have specified at least all required base attributes,

Reviewer 2: ```suggestion
    You can override this configuration, but make sure that all required base attributes are specified,
```

    otherwise it is not possible to create a new `CatalogElement` from the serialized version.

!!! tip

>     Also make sure that the attributes you have specified are either simple datatypes (int, float, boolean, string, array) or instances of [FieldInterface](../../../api/fields_for_ecommerce_data/fields_for_ecommerce_data.md).

Reviewer: "attributes that you specify"

    All other objects will be ignored, because it is not possible to assure that the serializing process will work properly.

## CatalogElementSerializationListener

> To store the `CatalogElement` in the basket line the Doctrine listener is used. This listener uses the serialize service.

Reviewer: The basket line uses the Doctrine listener to store the `CatalogElement`.

    `siso_basket.catalog_element_serialization_listener`

> This listener listens to different events to make sure that the catalog element is ALWAYS of type `CatalogElement` in the shop but `longtext` (serialized string) in the database.

Reviewer: Could it be:
"in the shop, and of `longtext` (serialized string) type in the database"?
